### ISSUE TITLE
Query processor API should handle race conditions better

### ISSUE THREAD

**Issue by attekei** [metabase/metabase#5044]
When writing E2E tests and running a query API call before a previous query API call finishes, I occasionally encounter an error message like this (and the server returns 400):
```
05-09 17:08:15 WARN metabase.query-processor :: Query failure: ERROR: duplicate key value violates unique constraint "pk_query"
  Detail: Key (query_hash)=(\xe4a5473033f26b71933be476787ed571cd1a9fd844c168e9fae878e1c87e128e) already exists.
["models.query$update_average_execution_time_BANG_.invokeStatic(query.clj:39)"
 "models.query$update_average_execution_time_BANG_.invoke(query.clj:28)"
 "query_processor$save_query_execution_BANG_.invokeStatic(query_processor.clj:112)"
 "query_processor$save_query_execution_BANG_.invoke(query_processor.clj:108)"
 "query_processor$save_and_return_successful_query_BANG_.invokeStatic(query_processor.clj:143)"
 "query_processor$save_and_return_successful_query_BANG_.invoke(query_processor.clj:133)"
 "query_processor$run_and_save_query_BANG_.invokeStatic(query_processor.clj:192)"
 "query_processor$run_and_save_query_BANG_.invoke(query_processor.clj:185)"
 "query_processor$eval37704$dataset_query__37709$fn__37710.invoke(query_processor.clj:223)"
 "query_processor$eval37704$dataset_query__37709.invoke(query_processor.clj:209)"
```

@camsaul explained that this is due to a race condition. We should probably either throw a more descriptive error message (like "this is probably a race condition") or handle the situation in some better way.

I don't know how common race conditions like this are in production instances. In E2E tests I'm able to circumvent this by adding extra safeguards that the previous query API call has been completed before starting a new one.

**Comment by camsaul** [comment]
this is not something that I think would be likely to happen in real life since multiple people would have to run the same brand-new query at the exact same millisecond to trigger this. Fixed nonetheless.

In general tho e2e tests should probably wait for API calls to finish before continuing to do whatever it is that they're doing or you're likely to run into other test failures (values not being updated when they're subsequently fetched, etc.)

**Comment by salsakran** [comment]
closed via #5045
